### PR TITLE
Coooooookie jar

### PR DESCRIPTION
**Pull request by kayladnls**
This PR:
- [x] Adds in CookieJar for cookie management

:cookie::cookie::cookie::cookie::cookie::cookie::cookie:

### CODE REVIEW

File: src/Request.php [hunk 1]
```diff
@@ -2,16 +2,26 @@
...
+        array $serverParams = [],
+        array $uploadedFiles = [],
+        $uri = null,
+        $method = null,
+        $body = 'php://input',
+        array $headers = []
+    ) {
+        $this->cookieJar = $jar ?: new CookieJar();
+
+        parent::__construct($serverParams, $uploadedFiles, $uri, $method, $body, $headers);
```
Comment (rcatlin) [inline src/Request.php]: Can we move the parent's constructor call as the first line of code in the constructor's body?

File: src/Piston.php [hunk 2]
```diff
@@ -75,6 +75,8 @@ public function launch()
         $this->response = $this->dispatch($this->request, $this->response);
         $this->response->compileContent();
 
+        $this->request = $this->request->fromCookieJar();
```
Comment (rcatlin) [inline src/Piston.php]: This doesn't seem right.

We get the request from the cookie jar?

Comment (kayladnls) [inline src/Piston.php]: in diactoros Requests are immutable. So, changing the cookie params actually returns a new instance of a Request object with those cookie params. 

I wasn't sure what to call this function, but essentially it takes the exiting cookie jar, creates a new request object, adds the cookie params to the new request and and returns that. 

Thoughts?

Comment (kayladnls) [inline src/Piston.php]: PS: As soon as this is merged I'm going to make another PR to make the other setter methods on Piston/Request immutable so it's more in line with Diactoros.

Comment (rcatlin) [inline src/Piston.php]: `$this->request->withCookieJar()` would make more sense. Thoughts?

Comment (rcatlin) [inline src/Piston.php]: `Request#withCookieJar` would also be more in line with Diactoros' naming convention.

`Request` also has:
* `Request#withRequestTarget()`
* `Request#withMethod()`
* `Request#withUri()`

**Comment by rcatlin** [comment]
:+1: Given you're okay with `fromCookieJar` or want to switch to `withCookieJar`

File: src/Piston.php (hunk 2 above)
Comment (kayladnls) [inline src/Piston.php]: I wasn't sure if `withCookieJar` would accurate. The Request object has a CookieJar inside it. The `fromCookieJar` method is transfering the contents of that CookieJar into the `cookieParameters` property.

**Comment by kayladnls** [comment]
@rcatlin take a quick gander now? I refactored so it regenerated the Request with each new cookie added.

**Comment by rcatlin** [comment]
:+1:
